syntax = "proto3";

package dataNode;

// Mensaje para solicitar la descarga de un bloque de archivo
message DownloadBlockRequest {
  string file_name = 1;
  string block_name = 2;
}

// Respuesta que contiene el bloque de archivo descargado
message DownloadBlockResponse {
  bytes block_data = 1;
}

// Mensaje para solicitar la carga de un nuevo bloque de archivo
message UploadBlockRequest {
  string file_name = 1;
  string block_name = 2;
  bytes block_data = 3;
}

// Respuesta que indica el éxito de la carga del bloque de archivo
message UploadBlockResponse {
  string message = 1;
}

// Definición del servicio gRPC para la transferencia de bloques de archivo
service DataNodeService {
  // Método para descargar un bloque de archivo
  rpc DownloadBlock(DownloadBlockRequest) returns (DownloadBlockResponse);

  // Método para cargar un nuevo bloque de archivo
  rpc UploadBlock(UploadBlockRequest) returns (UploadBlockResponse);
}

// Definición de la clase base para la implementación de los métodos del servicio DataNode
service DataNodeServicer {
  // Método para descargar un bloque de archivo
  rpc DownloadBlock(DownloadBlockRequest) returns (DownloadBlockResponse);

  // Método para cargar un nuevo bloque de archivo
  rpc UploadBlock(UploadBlockRequest) returns (UploadBlockResponse);
}

